### PR TITLE
Add cond_sub and sub_clamp operations to atomicrmw

### PR DESCRIPTION
**Pull request by anjenner**
These both perform conditional subtraction, returning the minuend and zero respectively, if the difference is negative.

AMDGPU has instructions for these. Currently we use target intrinsics for these, but those do not carry the ordering and syncscope. Add these to atomicrmw so we can carry these and benefit from the regular legalization processes. Drop and upgrade llvm.amdgcn.atomic.cond.sub/csub to atomicrmw.

Note that in GFX12 onwards "csub" is renamed to "sub_clamp" - the atomicrmw operation uses the newer name.

### CODE REVIEW

**Comment by tschuett** [comment]
Could you please extend 
https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/CodeGen/GlobalISel/MachineIRBuilder.h

Thanks.

File: llvm/test/CodeGen/X86/atomicrmw-cond-sub-clamp.ll [hunk 1]
```diff
@@ -0,0 +1,153 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -mtriple x86_64-pc-linux < %s | FileCheck %s
```
Comment (RKSimon) [inline llvm/test/CodeGen/X86/atomicrmw-cond-sub-clamp.ll:2]: Please 32-bit test coverage as well

File: llvm/lib/Target/AMDGPU/AMDGPUInstructions.td [hunk 2]
```diff
@@ -628,16 +628,16 @@ defm int_amdgcn_flat_atomic_fadd : global_addr_space_atomic_op;
...
 defm int_amdgcn_flat_atomic_fadd : local_addr_space_atomic_op;
 defm int_amdgcn_global_atomic_ordered_add_b64 : noret_op;
 defm int_amdgcn_flat_atomic_fmin_num : noret_op;
 defm int_amdgcn_flat_atomic_fmax_num : noret_op;
 defm int_amdgcn_global_atomic_fmin_num : noret_op;
 defm int_amdgcn_global_atomic_fmax_num : noret_op;
-defm int_amdgcn_atomic_cond_sub_u32 : local_addr_space_atomic_op;
-defm int_amdgcn_atomic_cond_sub_u32 : flat_addr_space_atomic_op;
-defm int_amdgcn_atomic_cond_sub_u32 : global_addr_space_atomic_op;
+//defm int_amdgcn_atomic_cond_sub_u32 : local_addr_space_atomic_op;
```
Comment (tschuett) [inline llvm/lib/Target/AMDGPU/AMDGPUInstructions.td:638]: What?

File: llvm/docs/LangRef.rst [hunk 3]
```diff
@@ -11259,6 +11261,8 @@ operation argument:
 -  fmin: ``*ptr = minnum(*ptr, val)`` (match the `llvm.minnum.*`` intrinsic)
 -  uinc_wrap: ``*ptr = (*ptr u>= val) ? 0 : (*ptr + 1)`` (increment value with wraparound to zero when incremented above input value)
 -  udec_wrap: ``*ptr = ((*ptr == 0) || (*ptr u> val)) ? val : (*ptr - 1)`` (decrement with wraparound to input value when decremented below zero).
+-  cond_sub: ``*ptr = (*ptr u>= val) ? *ptr - val : *ptr`` (subtract only if result would be positive).
+-  sub_clamp: ``*ptr = (*ptr u>= val) ? *ptr - val : 0`` (subtract with clamping of negative results to zero).
```
[the remark this answers is not in the document]
Comment (nikic) [reply llvm/docs/LangRef.rst:11265]: It looks like sub_clamp does the operation of the usub_sat intrinsic, so maybe call it usub_sat?

[the remark this answers is not in the document]
Comment (nikic) [reply llvm/docs/LangRef.rst:11265]: And maybe cond_sub should be usub_cond.

File: llvm/test/Bitcode/amdgcn-atomic.ll [hunk 4]
```diff
@@ -249,3 +249,84 @@ define <2 x i16> @upgrade_amdgcn_ds_fadd_v2bf16__missing_args_as_i16(ptr addrspa
...
+  ret void
+}
+
+declare i32 @llvm.amdgcn.atomic.cond.sub.i32.p1(ptr addrspace(1) nocapture, i32, i32 immarg, i32 immarg, i1 immarg) #0
+declare i32 @llvm.amdgcn.atomic.cond.sub.i32.p3(ptr addrspace(3) nocapture, i32, i32 immarg, i32 immarg, i1 immarg) #0
+declare i32 @llvm.amdgcn.atomic.cond.sub.i32.p0(ptr nocapture, i32, i32 immarg, i32 immarg, i1 immarg) #0
+
+declare i32 @llvm.amdgcn.atomic.csub.i32.p1(ptr addrspace(1) nocapture, i32, i32 immarg, i32 immarg, i1 immarg) #0
+declare i32 @llvm.amdgcn.atomic.csub.i32.p3(ptr addrspace(3) nocapture, i32, i32 immarg, i32 immarg, i1 immarg) #0
+declare i32 @llvm.amdgcn.atomic.csub.i32.p0(ptr nocapture, i32, i32 immarg, i32 immarg, i1 immarg) #0
```
Comment (arsenm) [inline llvm/test/Bitcode/amdgcn-atomic.ll:332]: Check an i64 case is handled correctly

**Comment by anjenner** [comment]
I have created a new pull request, https://github.com/llvm/llvm-project/pull/105553 , which addresses all the mentioned issues including splitting the commit into two.

**Pull request closed without merging**